sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs

```diff
@@ -262,6 +262,41 @@ protected virtual BlobBaseClient WithSnapshotImpl(string snapshot)
             return new BlobBaseClient(builder.Uri, Pipeline);
         }
 
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BlobClient"/> class
+        /// with an identical <see cref="Uri"/> source but the specified
+        /// <paramref name="customerProvidedKey"/> customer provided key.
+        /// </summary>
+        /// <param name="customerProvidedKey">
+        /// The customer provided key to be used by the service to encrypt data.
+        /// </param>
+        /// <returns>A new <see cref="BlobClient"/></returns>
+        public virtual BlobBaseClient WithCustomerProvidedKey(CustomerProvidedKey customerProvidedKey) =>
+            WithCustomerProvidedKeyImpl(customerProvidedKey);
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="BlobClient"/> class
+        /// with an identical <see cref="Uri"/> source but the specified
+        /// <paramref name="customerProvidedKey"/> customer provided key.
+        /// </summary>
+        /// <param name="customerProvidedKey">
+        /// The customer provided key to be used by the service to encrypt data.
+        /// </param>
+        /// <returns>A new <see cref="BlobClient"/></returns>
+        protected virtual BlobBaseClient WithCustomerProvidedKeyImpl(CustomerProvidedKey customerProvidedKey)
+        {
+            var uriBuilder = new UriBuilder(Uri)
+            {
+                Scheme = Constants.Blob.Https,
+                Port = Constants.Blob.HttpsPort
+            };
+            return new BlobBaseClient(
+                uriBuilder.Uri,
+                Pipeline,
+                new BlobClientOptions(customerProvidedKey: customerProvidedKey));
```

Comment:
I'm kind of worried about this pattern of passing BlobClientOptions but not using it to build the pipeline.  I think this should probably be changed to just pass the customer provided key.

We also shouldn't suffix customer visible methods with `Impl`.